sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1267,6 +1272,11 @@ class Analyzer(
           val newOutput = oldVersion.generatorOutput.map(_.newInstance())
           Seq((oldVersion, oldVersion.copy(generatorOutput = newOutput)))
 
+        case oldVersion: Union
+            if oldVersion.producedAttributes.intersect(conflictingAttributes).nonEmpty =>
+          val newOutput = oldVersion.unionOutput.get.map(_.newInstance())
+          Seq((oldVersion, oldVersion.copy(unionOutput = Some(newOutput))))
```

Comment:
NOTE: Since `Union` has `unionOutput` in its class fields, we need this entry for dedup as with `Generate`.